Add labels to CreatePipelineRun modal

ncskier opened the pull request:
# Changes
Fixes #323
This PR turns the Annotations component into a reusable KeyValueList
component. This KeyValueList component is used to manage annotations in
the SecretsModal, and it is used in the CreatePipelineRun modal to
manage labels.

Here's an image of the labels in the CreatePipelineRun modal (feedback is welcome 🙂):
![Screen Shot 2019-11-05 at 5 26 43 PM](https://user-images.githubusercontent.com/16961496/68251765-ce9b5b80-fff1-11e9-8ddd-0a2008ad59f7.png)

/cc @AlanGreene 
/cc @carlos-logro 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

carlos-logro commented on src/components/SecretsModal/SecretsModal.scss:
```diff
@@ -14,10 +14,6 @@ limitations under the License.
-    width: 30%;
-  }
-
   .bx--form-item,
```
I already showed you this working but for sake of book keeping.
Instead of importing the `KeyValues` css file here to overwrite the margin for the `.bx--form-item`, it is less of hassle if you change this style to: 
``` css
.bx--modal-content > form > .bx--form-item,
  .bx--dropdown__wrapper {
    margin-bottom: $spacing-03;
  }
```
With this style the only `.bx--form-item` elements that will be affected will only be those first order child elements inside the form tag. Meaning that the ones inside `keyvalues` div will not be changed.

carlos-logro commented on src/containers/CreatePipelineRun/CreatePipelineRun.js:
```diff
@@ -90,10 +91,26 @@ const initialResourcesState = resourceSpecs => {
   constructor(props) {
     super(props);
 
+    this.handleAddLabel = this.handleAddLabel.bind(this);
```
You can avoid having to bind `this` to each handler by defining them as:

``` Javascript
handlerName = (arguments) => {}
```

ncskier replied:
Oh, cool! I did not know this 👍

carlos-logro commented on src/containers/CreatePipelineRun/CreatePipelineRun.js:
```diff
@@ -172,6 +206,54 @@ class CreatePipelineRun extends React.Component {
+    this.setState(prevState => {
+      // Update labels
+      const labels = [...prevState.labels];
+      const l = labels.pop();
```
`removedLabel` makes more sense than just `l`

carlos-logro commented on src/containers/SecretsModal/SecretsModal.js:
```diff
@@ -59,15 +59,14 @@ export /* istanbul ignore next */ class SecretsModal extends Component {
-            .toString(36)
-            .substring(2, 11)
+          valuePlaceholder: 'https://github.com',
+          id: 'annotation0'
```
as per Alan Green's [comment](https://github.com/tektoncd/dashboard/pull/188#discussion_r295946500) on using index or array length for key value: 

> key should be a value derived from the data being rendered as react uses it to determine which elements need to be updated during the render cycle in response to data change, using the array index is not a good idea as it can lead to unexpected behaviour.

ncskier replied:
Ok... it was this before
```
Math.random()
    .toString(36)
    .substr(2, 9);
```
so I'll update the `id` to this in the SecretsModal and CreatePipelineRun modal.

carlos-logro commented on src/containers/SecretsModal/SecretsModal.js:
```diff
@@ -228,26 +212,29 @@ export /* istanbul ignore next */ class SecretsModal extends Component {
     const stateValue = e.selectedItem.id;
```
In this handler, for any annotation value that is blank or one of the docker or git default text, it will go ahead and change those value to the default one of the new selected access value (does not affect unique values). Which means that if it was previously invalid (because it was left blank) then it also removes this invalid designation since it changed it to a correct default value. That being said, I could not see this behavior when I pulled the PR. 


Example: 
Here we see that each field has a default git access value.
<img width="752" alt="Screen Shot 2019-11-06 at 4 38 03 PM" src="https://user-images.githubusercontent.com/49996607/68340168-e804dc00-00b3-11ea-9e82-4b6f8968e446.png">

Now, in this one I delete the second annotation value and we can see that it marks it as invalid because it is blank.
<img width="752" alt="Screen Shot 2019-11-06 at 4 38 12 PM" src="https://user-images.githubusercontent.com/49996607/68340212-f9e67f00-00b3-11ea-9ab7-c0a85f0a24a9.png">


Finally, I modified the last annotation value to be a unique one and then went ahead and change the `Access To` value from `Git` to `Docker`.  We can now see how after this, the second annotation value still is marked as invalid although it has a default docker access to value.
<img width="750" alt="Screen Shot 2019-11-06 at 4 38 26 PM" src="https://user-images.githubusercontent.com/49996607/68340252-0ec31280-00b4-11ea-9060-5f4d190f98bf.png">

ncskier replied:
Good catch @carlos-logro! The latest code I pushed fixes this error if you want to give it another try 👍

carlos-logro commented on src/containers/SecretsModal/SecretsModal.js:
```diff
@@ -59,15 +59,14 @@ export /* istanbul ignore next */ class SecretsModal extends Component {
       annotations: [
         {
-          label: `tekton.dev/git-0`,
+          key: 'tekton.dev/git-0',
```
what's the difference between `key` & `keyPlaceholder`?

ncskier replied:
The `key`, is the actual text that is in the "key" TextInput (the `value` prop), whereas the `keyPlaceholder` is the placeholder text in the "key" TextInput (the `placeholder` prop).

Do you think that these names are confusing/do you have an idea for an alternative name?

AlanGreene commented:
I'd like to get some design input on the controls for adding/removing items from the list, in particular for removal... IMO having the remove button at the top of the list remove an item from the bottom of the list seems odd. Since this is a pattern we're starting to reuse in multiple places in the UI we should make sure we're not causing confusion.
![image](https://user-images.githubusercontent.com/2829095/68393313-b3505d80-0163-11ea-9a2c-586d295c703d.png)

ncskier commented on src/components/KeyValueList/KeyValue.test.js:
```diff
@@ -0,0 +1,180 @@
+  const props = {
+    onChange() {},
+    legendText: 'Legend Text',
+    invalidFields: [],
```
Oops, I'll fix all of those 😄 good catch

ncskier commented on src/components/KeyValueList/index.js:
```diff
@@ -0,0 +1,14 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
```
Hm, that's an interesting idea. @AlanGreene what do you think about a "How to Use" comment?
I think the storybook sort of fills this role... which reminds me, I will add a storybook entry for the KeyValueList.

AlanGreene replied:
If this is a purely presentational UI component it should be in packages/components so it can be published and re-used via the @tektoncd/dashboard-components library with our other components.

Yes a story would be the best way to document this, and could provide a few different examples of use if there are interesting variations.

ncskier replied:
Ok, the directory structure has me all confused now 😅 it is a purely presentational UI component, so I will move it to `packages/components`

AlanGreene replied:
any components remaining in `src/components` should be updated to ensure they're pure and moved to `packages/components` so they can be used by others, looks like we're down to a few SCSS files, the Secrets components, TextInput and TooltipDropdown. I'll create an issue to track the remaining work

ncskier commented:
Thanks for the thorough feedback @carlos-logro & @AlanGreene 🙂 I have updated the code to addressed all of your comments.

> I'd like to get some design input on the controls for adding/removing items from the list, in particular for removal... IMO having the remove button at the top of the list remove an item from the bottom of the list seems odd. Since this is a pattern we're starting to reuse in multiple places in the UI we should make sure we're not causing confusion.

@AlanGreene you make a good point about the remove behavior being a little unintuitive. Fortunately, since the KeyValueList is now a reusable component, it will be relatively easy to modify 👍 Since we already have this component in the Dashboard, would it be alright if we handle this redesign under a separate issue (separate from this PR)? I think that it could get messy redesigning the component under this PR, because this PR is primarily concerned with adding labels to the Create PipelineRun modal.

AlanGreene commented:
Hey @ncskier, yes let's create an issue to make sure that's tracked and try to get to it soon. I'll take another looks at your changes this evening.

AlanGreene commented on src/containers/CreatePipelineRun/CreatePipelineRun.js:
```diff
@@ -367,6 +453,20 @@ class CreatePipelineRun extends React.Component {
           )}
+          <FormGroup legendText="">
+            <KeyValueList
+              legendText="Labels"
```
strings should be externalised using react-intl (`intl.formatMessage(...)` or the `<FormattedMessage ... />` component)

AlanGreene commented on src/containers/CreatePipelineRun/CreatePipelineRun.js:
```diff
@@ -367,6 +453,20 @@ class CreatePipelineRun extends React.Component {
               />
             </FormGroup>
           )}
+          <FormGroup legendText="">
```
should the legend text be applied to the FormGroup instead or can we remove the group here?

ncskier replied:
The current style has the legend text inline with the Add & Remove buttons. We can't do this with a normal FormGroup legendText. So, instead, I'm using the styles that were in the SecretModal which make it look like a FormGroup's legendText.

AlanGreene commented on src/containers/SecretsModal/SecretsModal.js:
```diff
@@ -376,12 +370,18 @@ export /* istanbul ignore next */ class SecretsModal extends Component {
-            handleRemove={this.handleRemove}
-            handleAdd={this.handleAdd}
+          <KeyValueList
+            legendText="Server URL"
```
react-intl

AlanGreene commented on src/containers/SecretsModal/SecretsModal.js:
```diff
@@ -376,12 +370,18 @@ export /* istanbul ignore next */ class SecretsModal extends Component {
+            legendText="Server URL"
+            invalidText="Server URL required."
+            ariaLabelKey="This is the tag Tekton uses for its resources."
+            ariaLabelValue="This is the url for the given Tekton resource."
```
URL

AlanGreene commented on src/containers/CreatePipelineRun/CreatePipelineRun.js:
```diff
@@ -367,6 +453,20 @@ class CreatePipelineRun extends React.Component {
+          <FormGroup legendText="">
+            <KeyValueList
+              legendText="Labels"
+              invalidText="Labels must follow the kubernetes labels syntax"
```
Should we provide a link to documentation here or are users expected to know the rules?

ncskier replied:
I can add a link to https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set. Since I'm passing this as the prop `invalidText`, and this string is embedded in `<p>{invalidText}</p>`, I don't know how to embed a link in the text...
Will I have to pass `invalidText={<p>...<a href="">...</a></p>}` and then refer to that within the KeyValueList as `{invalidText}`?

ncskier replied:
I've updated the text to have a link 👍

ncskier commented on packages/components/src/components/KeyValueList/KeyValue.test.js:
```diff
@@ -0,0 +1,194 @@
+  fireEvent.click(removeButton);
+
+  expect(props.onAdd).toHaveBeenCalledTimes(5);
+  expect(props.onRemove).toHaveBeenCalledTimes(3);
```
With the current implementation, we don't care about the values because you add/remove to/from the list of KeyValues like it's a stack. There is no index or important information being passed in the `onAdd` or `onRemove` callback functions.

ncskier commented on packages/components/src/components/KeyValueList/KeyValueList.js:
```diff
@@ -0,0 +1,110 @@
+          {deleteButton && value && (
+            <Delete
+              className="deleteInputIcon"
+              aria-label={`Delete Button#${index}. Deletes entire entry in current text field.`}
```
@AlanGreene I am not familiar with `react-intl`. Would I wrap the `Delete` and `Add` components in an `IntlProvider`?

AlanGreene replied:
The `IntlProvider` is already present higher up the tree so no need to worry about that.

You'll need to make sure the component is wrapped with the `injectIntl` higher order component: `export default injectIntl(KeyValueList);`, this makes the `intl` prop available.

Then you'll need to construct the string something like this:

```js
const deleteAriaLabel = this.props.intl.formatMessage(
  {
    id: 'dashboard.keyValueList.deleteButtonLabel',
    defaultValue: 'Delete Button#{index}. Deletes entire entry in current text field.' // NOTE: there's no '$' before the `index` placeholder
  },
  { index }
);
```

Then you'd run `npm run i18n:extract` to update the message bundle with the new string and include this in your changes.

Do we need to include the index in the message? Users should already have the context of the current row, screen readers should also already announce this as a button so no need to repeat it in the label.

With these changes the string becomes:

```js
const deleteAriaLabel = this.props.intl.formatMessage(
  {
    id: 'dashboard.keyValueList.deleteButtonLabel',
    defaultValue: 'Delete entire entry in current text field.'
  }
);
```

ncskier commented on packages/components/src/components/KeyValueList/KeyValueList.js:
```diff
@@ -0,0 +1,110 @@
+              : 'removeIcon'
+          }
+          onClick={onRemove}
+          aria-label={`Remove Button. Removes ${legendText} entry.`}
```
Sorry, I'm not sure what you mean by this. What would I do differently if it was useful or wasn't useful for all users?

AlanGreene replied:
`aria-label` will only be read by screen readers or similar assistive technologies. If the text in this label is useful / intended for all users it should be a tooltip, an actual label, or some other visible element on the page.

ncskier replied:
Ah, gotcha, thanks for clarifying. I personally don't think that the text will help users without screen readers. I think that the icons on the buttons are pretty self-explanatory.

ncskier commented on packages/utils/src/utils/index.js:
```diff
@@ -169,3 +169,10 @@ export function updateUnexecutedSteps(steps) {
 }
+
+// Generates a unique id
+export function generateId() {
```
Yes, it used the index in the list before, but I received [this comment](https://github.com/tektoncd/dashboard/pull/699#discussion_r343336579), so I changed it to the `generateId()` function.

ncskier replied:
I'm a little confused about this, but I added a prefix for each id based on the item in the KeyValueList; for example:
```javascript
id: generateId(`annotation${annotations.length}-`)
```
Do you think this is alright?

ncskier commented:
Thanks for the review @AlanGreene! I think that I've addressed all of your comments. Can you please take another look when you get the chance?

AlanGreene commented on packages/components/src/components/KeyValueList/KeyValueList.js:
```diff
@@ -0,0 +1,143 @@
+            { legendText }
+          )}
+        />
+        <Add
```
Sorry I didn't pick up on this in previous reviews, I didn't realise these were just icons and not buttons.
We should use Button components here with the icon prop set, this will ensure keyboard accessibility etc. works as expected. If you want to hide the button border and background you can set `kind="ghosh"`. See LabelFilter, LogoutButton, or Rerun for examples.

AlanGreene commented on src/containers/CreatePipelineRun/CreatePipelineRun.js:
```diff
@@ -367,6 +454,50 @@ class CreatePipelineRun extends React.Component {
+                defaultMessage: 'Labels'
+              })}
+              invalidText={intl.formatMessage(
+                {
```
We can't translate fragments of a sentence in isolation (e.g. 'kubernetes labels syntax'), it won't work for many languages.
Instead what we want is to add placeholders for the link markup so the sentence can be translated as a whole, with the correct context for each part.

Something like this:
'Labels must follow the {0}kubernetes labels syntax{1}.'

ncskier replied:
@AlanGreene I don't think it lets me split up JSX tags like that:
![Screen Shot 2019-12-05 at 2 02 15 PM](https://user-images.githubusercontent.com/16961496/70265722-f99bdb00-1768-11ea-94fa-234802c277d4.png)

ncskier replied:
I could maybe do something like this to use an icon instead of text as the link, so we won't have to translate it?
![Screen Shot 2019-12-05 at 2 11 32 PM](https://user-images.githubusercontent.com/16961496/70265802-28b24c80-1769-11ea-81c2-ab73c2a8bf15.png)

AlanGreene replied:
ok looks like react-intl decided this was something people don't need, but there's still a way to get what we want. React's scarily named `dangerouslySetInnerHTML` to the rescue 😆 In this case it's safe as we fully control the content, there's no user data involved, we can consider the HTML content sanitised.

![image](https://user-images.githubusercontent.com/2829095/70322058-f6642600-1820-11ea-95a3-0e450a50f336.png)

```js
invalidText={
  <span
    dangerouslySetInnerHTML /* eslint-disable-line react/no-danger */={{
      __html: intl.formatMessage(
        {
          id: 'dashboard.createPipelineRun.labels.invalidText',
          defaultMessage:
            'Labels must follow the {0}kubernetes labels syntax{1}.'
        },
        [
          `<a
              href="https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set"
              target="_blank"
              rel="noopener noreferrer"
            >`,
          '</a>'
        ]
      )
    }}
  />
}
```

ncskier replied:
Awesome, thanks Alan! I just made the change.

AlanGreene commented on src/nls/messages_en.json:
```diff
@@ -17,6 +22,11 @@
     "dashboard.errorBoundary.pageError": "Error loading page",
     "dashboard.header.logOut": "Log out",
+    "dashboard.keyValueList.addButton.ariaLabel": "Add Button. Adds {legendText} entry.",
+    "dashboard.keyValueList.deleteButton.ariaLabel": "Delete Button#{index}. Deletes entire entry in current text field.",
```
I'm still struggling a bit with these strings.

'Delete Button#0', will this make sense to a user?
'current text field', how do I know which is the 'current' text field?

ncskier replied:
You're right, those strings don't really make sense. Since there's already text on the screen for each of these elements, I'm going to remove the aria labels. Is that alright?
